pkg/i2gw/provider.go

```diff
@@ -0,0 +1,107 @@
+
+// ProviderName is a string alias that stores the concrete Provider name.
+type ProviderName string
+
+// ProviderConstructor is a construction function that constructs concrete
+// implementations of the Provider interface.
+type ProviderConstructor func(conf *ProviderConf) Provider
+
+// ProviderConf contains all the configuration required for every concrete
+// Provider implementation.
+type ProviderConf struct{}
+
+// The Provider interface specifies the required functionality which needs to be
+// implemented by every concrete Ingress/Gateway-API provider, in order for it to
+// be used.
+type Provider interface {
+	CustomResourceReader
+	ResourceConverter
+}
+
+type CustomResourceReader interface {
+
+	// ReadResourcesFromCluster reads custom resources associated with
+	// the underlying Provider implementation from the kubernetes cluster.
+	ReadResourcesFromCluster(ctx context.Context, customResources interface{}) error
+
+	// ReadResourcesFromFiles reads custom resources associated with
+	// the underlying Provider implementation from the files.
+	ReadResourcesFromFiles(ctx context.Context, customResources interface{}, filename string) error
+}
+
+// The ResourceConverter interface specifies all the implemented Gateway API resource
+// conversion functions.
+type ResourceConverter interface {
+
+	// IngressToGateway converts the received IngressResources associated
+	// with the Provider into GatewayResources.
+	IngressToGateway(resources IngressResources) (GatewayResources, field.ErrorList)
+}
+
+// IngressResources contains all Ingress related objects, and Provider specific
+// custom resources.
+type IngressResources struct {
+	Ingresses       []networkingv1.Ingress
+	CustomResources interface{}
+}
+
+// GatewayResources contains all Gateway-API objects.
+type GatewayResources struct {
+	Gateways   map[GatewayKey]gatewayv1beta1.Gateway
+	HTTPRoutes map[HTTPRouteKey]gatewayv1beta1.HTTPRoute
+}
+
+// GatewayKey is a unique identifier for a gateway object.
+// Constructed by namespace:name.
+type GatewayKey string
+
+// GatewayToGatewayKey assembles the GatewayKey out of a Gateway.
+func GatewayToGatewayKey(g gatewayv1beta1.Gateway) GatewayKey {
+	return GatewayKey(g.Namespace + ":" + g.Name)
+}
+
+// HTTPRouteKey is a unique identifier for an HTTPRoute object.
+// Constructed by namespace:name.
+type HTTPRouteKey string
+
+// HTTPRouteToHTTPRouteKey assembles the HTTPRouteKey out of an HTTPRoute.
+func HTTPRouteToHTTPRouteKey(r gatewayv1beta1.HTTPRoute) HTTPRouteKey {
+	return HTTPRouteKey(r.Namespace + ":" + r.Name)
+}
+
+// FeatureParser is a function that reads the IngressResources, and applies
+// the appropriate modifications to the GatewayResources.
+//
+// Different FeatureParsers will run in undetermined order. The function must
+// modify / create only the required fields of the gateway resources and nothing else.
+type FeatureParser func(IngressResources, *GatewayResources) field.ErrorList
```

Comment:
Maybe we should think of a different name to `IngressResources` its a bit confusing. 
Essentially since it holds both ingresses and custom resources it is more of "resources we are going to transform"

Maybe "ToGatewayResources" ? Not sure I like that either

Reply:
The best next thing I can think about is `IngressRelatedResources` and `GatewayRelatedResources`, but in my opinion, they are unnecessarily long.
I don't really see a problem with the current naming convention, but if you or others have a better suggestion that you prefer, please let me know

Reply:
Made the switch to `InputResources` I think it's descriptive enough.
I mark it as resolved but please reopen this discussion if you want to further debate on it.